packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts

```diff
@@ -0,0 +1,242 @@
+        // method is not called.
+        this.references = cloneFluidObjectReferences(summarizeResult.references);
+        return summarizeResult;
+    }
+
+    /**
+     * Override the createRoot method to return an instance of SummarizerNodeWithReferences.
+     */
+    public static createRoot(
+        logger: ITelemetryLogger,
+        /** Summarize function */
+        summarizeInternalFn: (fullTree: boolean, trackState: boolean) => Promise<ISummarizeInternalResult>,
+        /** Sequence number of latest change to new node/subtree */
+        changeSequenceNumber: number,
+        /**
+         * Reference sequence number of last acked summary,
+         * or undefined if not loaded from summary.
+         */
+        referenceSequenceNumber: number | undefined,
+        config: ISummarizerNodeConfig = {},
+        /** Function to get initial Fluid object references */
+        getInitialFluidObjectReferencesFn?: () => Promise<IFluidObjectReferences[]>,
+    ): SummarizerNodeWithReferences {
+        const maybeSummaryNode = referenceSequenceNumber === undefined ? undefined : new SummaryNode({
+            referenceSequenceNumber,
+            basePath: undefined,
+            localPath: EscapedPath.create(""), // root hard-coded to ""
+        });
+        return new SummarizerNodeWithReferences(
+            logger,
+            summarizeInternalFn,
+            config,
+            changeSequenceNumber,
+            maybeSummaryNode,
+            undefined /* initialSummary */,
+            getInitialFluidObjectReferencesFn,
+        );
+    }
+
+    /**
+     * Override the createChild method to return an instance of SummarizerNodeWithReferences.
+     */
+    public createChild(
+        /** Summarize function */
+        summarizeInternalFn: (fullTree: boolean, trackState: boolean) => Promise<ISummarizeInternalResult>,
+        /** Initial id or path part of this node */
+        id: string,
+        /**
+         * Information needed to create the node.
+         * If it is from a base summary, it will assert that a summary has been seen.
+         * Attach information if it is created from an attach op.
+         */
+        createParam: CreateChildSummarizerNodeParam,
+        config: ISummarizerNodeConfig = {},
+        /** Function to get initial Fluid object references */
+        getInitialFluidObjectReferencesFn?: () => Promise<IFluidObjectReferences[]>,
+    ): ISummarizerNodeWithReferences {
+        assert(!this.children.has(id), "Create SummarizerNode child already exists");
+
+        const latestSummary = this.latestSummary;
+        let child: SummarizerNodeWithReferences;
+        switch (createParam.type) {
+            case CreateSummarizerNodeSource.FromAttach: {
+                let summaryNode: SummaryNode | undefined;
+                let initialSummary: IInitialSummary | undefined;
+                if (
```

Comment:
I only briefly looked through this code, so maybe my question is not valid, but still:
Any chance we can not have in this class logic around sequence numbers? I.e. can base class do all the heavy lifting and provide right accessors, and this object only be responsive for reading / writing GC info?
I worry about always increasing complexity of summarizing domain, and logic being spread across so many layers, that when things go wrong, we are left completely clueless on how to approach it.

Reply:
I totally agree. The thing is SummarizerNode is not designed to be inherited :-). So, I will have to make changes to it such that it can be inherited without requiring leaking the core logic. I will look into it.

Reply:
Restructured the code a little to do this. Can you please take a look and let me know.